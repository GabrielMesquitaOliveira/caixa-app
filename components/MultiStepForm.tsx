import { FormData } from '@/types';
import { yupResolver } from '@hookform/resolvers/yup';
import React, { useState } from 'react';
import { useForm } from 'react-hook-form';
import { Text, View } from 'react-native';
import * as yup from 'yup';
import StepOne from './StepOne';
import StepThree from './StepThree';
import StepTwo from './StepTwo';

interface MultiStepFormProps {
  onSubmit: (data: FormData) => void;
  onCancel: () => void;
}

// Schema de validação para cada passo
const validationSchemas = {
  stepOne: yup.object().shape({
    valorContratado: yup
      .number()
      .required('Valor é obrigatório')
      .min(1000, 'Valor mínimo é R$ 1.000')
      .max(1000000, 'Valor máximo é R$ 1.000.000'),
    prazoContratado: yup
      .number()
      .required('Prazo é obrigatório')
      .min(1, 'Prazo mínimo é 1 mês')
      .max(360, 'Prazo máximo é 360 meses'),
  }),
  stepTwo: yup.object().shape({
    sistemaAmortizacao: yup
      .string()
      .required('Sistema de amortização é obrigatório')
      .oneOf(['PRICE', 'SAC'], 'Selecione um sistema válido'),
    dataVencimento: yup
      .string()
      .required('Data de vencimento é obrigatória'),
  }),
  stepThree: yup.object().shape({
    nomePersonalizado: yup
      .string()
      .required('Nome personalizado é obrigatório')
      .min(3, 'Mínimo 3 caracteres')
      .max(50, 'Máximo 50 caracteres'),
    observacoes: yup
      .string()
      .max(500, 'Máximo 500 caracteres'),
  }),
};

const MultiStepForm: React.FC<MultiStepFormProps> = ({ onSubmit, onCancel }) => {
  const [step, setStep] = useState(1);
  const [formData, setFormData] = useState<FormData>({});

  // Hook forms separados para cada passo
  const stepOneForm = useForm({
    resolver: yupResolver(validationSchemas.stepOne),
  });

  const stepTwoForm = useForm({
    resolver: yupResolver(validationSchemas.stepTwo),
  });

  const stepThreeForm = useForm({
    resolver: yupResolver(validationSchemas.stepThree),
  });

  const onNextStep = (data: Partial<FormData>) => {
    setFormData({ ...formData, ...data });
    if (step < 3) {
      setStep(step + 1);
    } else {
      onSubmit({ ...formData, ...data } as FormData);
    }
  };

  const onPrevious = () => {
    if (step > 1) {
      setStep(step - 1);
    }
  };

  return (
    <View className="p-4">
      {/* Indicador de progresso */}
      <View className="flex-row justify-center mb-6">
        {[1, 2, 3].map((stepNumber) => (
          <View
            key={stepNumber}
            className={`w-8 h-8 rounded-full mx-2 flex items-center justify-center ${
              stepNumber <= step ? 'bg-blue-500' : 'bg-gray-300'
            }`}
          >
            <Text className="text-white font-bold">{stepNumber}</Text>
          </View>
        ))}
      </View>

      {/* Formulário atual */}
      {step === 1 && (
        <StepOne
          control={stepOneForm.control as any}
          errors={stepOneForm.formState.errors}
          handleSubmit={stepOneForm.handleSubmit}
          onNextStep={onNextStep}
          onCancel={onCancel}
        />
      )}
      {step === 2 && (
        <StepTwo
          control={stepTwoForm.control as any}
          errors={stepTwoForm.formState.errors}
          handleSubmit={stepTwoForm.handleSubmit}
          onNextStep={onNextStep}
          onPrevious={onPrevious}
          onCancel={onCancel}
          formData={formData}
        />
      )}
      {step === 3 && (
        <StepThree
          control={stepThreeForm.control as any}
          errors={stepThreeForm.formState.errors}
          handleSubmit={stepThreeForm.handleSubmit}
          onNextStep={onNextStep}
          onPrevious={onPrevious}
          onCancel={onCancel}
          formData={formData}
        />
      )}
    </View>
  );
};

export default MultiStepForm;
