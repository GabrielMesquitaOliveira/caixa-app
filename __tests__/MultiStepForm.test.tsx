import { render } from '@testing-library/react-native';
import React from 'react';
import MultiStepForm from '../components/MultiStepForm';

describe('MultiStepForm Component', () => {
  const mockOnSubmit = jest.fn();
  const mockOnCancel = jest.fn();

  it('renders correctly', () => {
    const { getByText } = render(<MultiStepForm onSubmit={mockOnSubmit} onCancel={mockOnCancel} />);
    
    expect(getByText('1')).toBeTruthy(); // Check if step indicator is rendered
  });

  it('matches snapshot', () => {
    const { toJSON } = render(<MultiStepForm onSubmit={mockOnSubmit} onCancel={mockOnCancel} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
