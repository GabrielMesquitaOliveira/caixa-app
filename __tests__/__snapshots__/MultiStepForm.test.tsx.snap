// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiStepForm Component matches snapshot 1`] = `
<View
  className="p-4"
>
  <View
    className="flex-row justify-center mb-6"
  >
    <View
      className="w-8 h-8 rounded-full mx-2 flex items-center justify-center bg-blue-500"
    >
      <Text
        className="text-white font-bold"
      >
        1
      </Text>
    </View>
    <View
      className="w-8 h-8 rounded-full mx-2 flex items-center justify-center bg-gray-300"
    >
      <Text
        className="text-white font-bold"
      >
        2
      </Text>
    </View>
    <View
      className="w-8 h-8 rounded-full mx-2 flex items-center justify-center bg-gray-300"
    >
      <Text
        className="text-white font-bold"
      >
        3
      </Text>
    </View>
  </View>
  <View
    className="flex-1 justify-center items-center"
  >
    <ActivityIndicator
      color="#007AFF"
      size="large"
    />
    <Text
      className="mt-4 text-gray-600"
    >
      Carregando modalidades...
    </Text>
  </View>
</View>
`;
